File: 2236-maximum-twin-sum-of-a-linked-list/maximum-twin-sum-of-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        
        while(fast.next.next!=null){
            slow = slow.next;
            fast = fast.next.next;
        }
        slow.next = reverse(slow.next);

        slow = head;
        fast = head;
        while(fast!=null  && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        int maxSum = 0;
        while(slow !=null){
            int currSum = head.val + slow.val;
            if(currSum > maxSum) maxSum = currSum;
            slow = slow.next;
            head = head.next;
        }
        return maxSum;
    }
    public ListNode reverse(ListNode curr){
        ListNode prev = null;
        while(curr!=null){
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }
}

File: 2236-maximum-twin-sum-of-a-linked-list/maximum-twin-sum-of-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middle(ListNode* head){
        ListNode* slow = head;
        ListNode* fast = head;

        while(fast != NULL){
            fast = fast -> next;
            if(fast != NULL){
                fast = fast -> next;
                slow = slow -> next;
            }
        }
        return slow;
    }

    ListNode* reverse(ListNode* head){
        ListNode* prev = NULL;
        ListNode* curr = head;

        while(curr != NULL){
            ListNode* fwd = curr -> next;
            curr->next = prev;
            prev = curr;
            curr = fwd;
        }
        return prev;
    }
    int pairSum(ListNode* head) {
        ListNode* middleNode = middle(head);
        middleNode = reverse(middleNode);
        int maxSum = INT_MIN;

        while(middleNode != NULL){
            maxSum = max(maxSum,(head->val + middleNode->val));
            head = head->next;
            middleNode = middleNode->next;
        }
        return maxSum;
    }
};